language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+/// The type of an edge
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+/// A borrow graph is represented as a map from a source nonce to the set of all edges
+/// coming out of it.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph<T>(BTreeMap<Nonce, BTreeSet<Edge<T>>>);
+
+impl<T> BorrowGraph<T>
+where
+    T: Ord,
+    T: Clone,
+{
+    /// creates a new empty borrow graph
+    pub fn new() -> Self {
+        BorrowGraph(BTreeMap::new())
+    }
+
+    /// adds a fresh nonce
+    pub fn add_nonce(&mut self, nonce: Nonce) {
+        checked_precondition!(!self.0.contains_key(&nonce));
+        self.0.insert(nonce, BTreeSet::new());
+    }
+
+    /// adds a weak edge
+    pub fn add_weak_edge(&mut self, from: Nonce, label: Label<T>, to: Nonce) {
+        checked_precondition!(self.0.contains_key(&from));
+        checked_precondition!(self.0.contains_key(&to));
+        checked_precondition!(self.0[&to].is_empty());
+        let new_edge = Edge {
+            edge_type: EdgeType::Weak,
+            label,
+            to,
+        };
+        self.0.get_mut(&from).unwrap().insert(new_edge);
+    }
+
+    /// adds a strong edge and factors other edges coming out of `from` with respect to the new edge
+    pub fn add_strong_edge(&mut self, from: Nonce, label: Label<T>, to: Nonce) {
+        checked_precondition!(self.0.contains_key(&from));
+        checked_precondition!(self.0.contains_key(&to));
+        checked_precondition!(self.0[&to].is_empty());
+        checked_precondition!(label.len() <= 1);
+
+        let new_edge = Edge {
+            edge_type: EdgeType::Strong,
+            label: label.clone(),
+            to,
+        };
+        self.0.remove(&to).unwrap();
+        if label.is_empty() {
+            let from_edge_set = self.0.remove(&from).unwrap();
+            let mut new_from_edge_set = BTreeSet::new();
+            new_from_edge_set.insert(new_edge);
+            self.0.insert(from, new_from_edge_set);
+            self.0.insert(to, from_edge_set);
+        } else {
+            let mut edge_set: Vec<Edge<T>> = self.0[&from]
+                .iter()
+                .filter(|x| !x.label.is_empty() && x.label[0] == label[0])
+                .cloned()
+                .collect();
+            let from_edge_set = self.0.get_mut(&from).unwrap();
+            edge_set.iter().for_each(|x| {
+                from_edge_set.remove(x);
+            });
+            from_edge_set.insert(new_edge);
+            edge_set.iter_mut().for_each(|x| {
+                x.label.remove(0);
+            });
```

Comment:
Do you think something like this could work instead?
```
    let from_edge_set = self.0.get_mut(&from).unwrap();

    let edge_set = self.0[&from]
        .iter_mut()
        .filter(|x| !x.label.is_empty() && x.label[0] == label[0])
        .for_each(|x| {
            from_edge_set.remove(x);
            x.label.remove(0);
        })
        .cloned()
        .collect();
    from_edge_set.insert(new_edge);
    self.0.insert(to, edge_set.into_iter().collect());
```

Reply:
I don't think you need the second `edge_set.into_iter().collect()`

Reply:
I tried it but ran into miscellaneous borrow issues.

Reply:
I have made an attempt to refactor this code.